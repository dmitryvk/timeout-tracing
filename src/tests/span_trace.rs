use std::time::Duration;

use serial_test::serial;
use tokio::{join, time::sleep};
use tracing::{info, instrument};
use tracing_error::ErrorLayer;
use tracing_subscriber::layer::SubscriberExt;

use crate::{TimeoutElapsed, tests::insta_trace_filters, timeout, trace::CaptureSpanTrace};

#[tokio::test]
#[serial]
async fn format_span_trace() {
    let subscriber = tracing_subscriber::registry().with(ErrorLayer::default());

    let _guard = tracing::subscriber::set_default(subscriber);
    let result = timeout(Duration::from_millis(100), CaptureSpanTrace, do_sleep()).await;

    assert!(matches!(result, Err(TimeoutElapsed { .. })));
    let mut err = result.err().unwrap();
    err.active_traces.sort_by_cached_key(ToString::to_string);
    insta::with_settings!({
        filters => insta_trace_filters()
    }, {
        insta::assert_debug_snapshot!(err);
        insta::assert_snapshot!(err);
    });
}

#[instrument]
async fn do_sleep() {
    info!("sleep before");
    join!(do_sleep_a(), do_sleep_b());
}

#[instrument]
async fn do_sleep_a() {
    info!("sleep a");
    sleep(Duration::from_secs(1)).await;
}

#[instrument]
async fn do_sleep_b() {
    info!("sleep b");
    sleep(Duration::from_secs(1)).await;
}
