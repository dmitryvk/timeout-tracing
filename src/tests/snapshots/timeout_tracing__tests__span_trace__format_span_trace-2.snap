---
source: src/tests/span_trace.rs
expression: err
---
timeout elapsed at:
trace 0:
   0: timeout_tracing::tests::span_trace::do_sleep_a
             at src/tests/span_trace.rs:[NNN]
   1: timeout_tracing::tests::span_trace::do_sleep
             at src/tests/span_trace.rs:[NNN]
trace 1:
   0: timeout_tracing::tests::span_trace::do_sleep_b
             at src/tests/span_trace.rs:[NNN]
   1: timeout_tracing::tests::span_trace::do_sleep
             at src/tests/span_trace.rs:[NNN]
